### ISSUE TITLE
TLS config issues

### ISSUE THREAD

**Issue by bcebere** [coredns/coredns#2935]
Hi

The TLS plugin overwrites the tls.Config from Caddy, and zeroes the minimum version, maximum version and default ciphersuites.

This results in some issues when using CoreDNS with TLS support:
 - SSLv3, TLSv1.0 and TLSv1.1 are accepted. 
 - The are some weak ciphersuites accepted.
 - No server preferred ciphersuite order is provided.
 - No TLS 1.3 support

Given that DNS-over-TLS and DNS-over-HTTPS are relatively new, I think it shouldn't be backwards compatible with those weak configs.

Some tools to reproduce the issues:
https://github.com/drwetter/testssl.sh/
or 
https://github.com/ssllabs/ssllabs-scan

Run them against a TLS-enabled CoreDNS to reproduce the issues.

I generated a patch here https://gist.github.com/bcebere/cc12acb8ecbb7527061ecade69b1619b

I tried several times to push a branch and create a pull request in CoreDNS, but I guess you have some restrictions and it's not possible. It didn't happen with other repos.

So, I'll leave the patch here,maybe it's useful.

**Comment by sharmaansh21** [comment]
@bcebere You need to fork the repo and push it to your fork and create a branch. I am happy to do that for you if you don't have time or do not wish to do it.
About the solution, I don't have much expertise in this area so maybe @chrisohaver @miekg can help here.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "[coredns/coredns] TLS config issues..." ]
>I generated a patch here https://gist.github.com/bcebere/cc12acb8ecbb7527061ecade69b1619b

Some comments on that patch:

`const pluginName = "tls"` is an indirection that doesn't make things more 
readable, just use "tls" directly as it now.

For the `hardenTLS` function is there anything from the std lib we can use? How 
do we know this is a good config?

Do we need to do something server side? https://github.com/coredns/coredns/blob/master/core/dnsserver/server_https.go

**Comment by bcebere** [comment]
I created a pull request now, thank you for the clarification @sharmaanshul21 
https://github.com/coredns/coredns/pull/2938


@miekg :
I removed the indirection, thank you for the insight.

I don't think there is some standard function, but if the tls plugin doesn't overwrite the default Caddy config, at least TLS 1.1 or earlier wouldn't be supported.

Second, I hardcoded some modern, strong ciphersuites. This is optional, I guess, but by default, Caddy uses some ciphersuites considered weak by now. - mostly for backwards compatibility with browsers, and I don't think is the use case in CoreDNS.

I can put here the results from one of the TLS testing tools, before and after, to see the differences yourself.


Server side, for the HTTPS one, there is no additional fix needed. If you want a clean HTTPS setup, maybe adding `Strict-Transport-Security` header would be nice. But it's optional, it's just for browsers to remember to use HTTPS everytime. not an actual use case for CoreDNS, I guess.

**Comment by bcebere** [comment]
I run a test with https://github.com/drwetter/testssl.sh/, before and after the patch.

TLDR, before the patch, the tool issues the following warnings(this is the actual output):

1. SSLv3      offered (NOT ok)
2. TLS 1.3    not offered
    TLS 1      offered
    TLS 1.1    offered
3. Triple DES Ciphers / IDEA                     offered (NOT ok)
4. Has server cipher order?     nope (NOT ok)
5. POODLE, SSL (CVE-2014-3566)               VULNERABLE (NOT ok)
6. BEAST (CVE-2011-3389)                     
             SSL3: AES256-SHA AES128-SHA DES-CBC3-SHA 
             TLS1: ECDHE-RSA-AES256-SHA AES256-SHA ECDHE-RSA-AES128-SHA AES128-SHA ECDHE-RSA-DES-CBC3-SHA DES-CBC3-SHA
7. LUCKY13 (CVE-2013-0169), experimental     potentially VULNERABLE


After the patch, there is no warning.
Full results attached

[before.txt](https://github.com/coredns/coredns/files/3337994/before.txt)
[after.txt](https://github.com/coredns/coredns/files/3337993/after.txt)

**Comment by miekg** [comment]
I think this is a good idea. However I'm not qualified, not have the time
to become one, to keep this up to date.

If you have github trouble, just mail the patch, I'll apply locally and
merge.

**Comment by bcebere** [comment]
> I think this is a good idea. However I'm not qualified, not have the time to become one, to keep this up to date. If you have github trouble, just mail the patch, I'll apply locally and merge.
> […](#)

As I mentioned above, I fixed my issues and I opened a pull request. Hope it helps.

https://github.com/coredns/coredns/pull/2938

**Comment by bcebere** [comment]
fixed in https://github.com/coredns/coredns/commit/6e7a5f56779c4ff1b317a98c5c8a1e9b472cab15
